Harden data and config protection .htaccess

MichaIng opened the pull request:
+ Set "`Satisfy All`" whenever available, as well on Apache 2.4+. This is required to override possible "`Satisfy Any`" on parent dir, which otherwise would allow public access, regardless of "[Require](https://httpd.apache.org/docs/2.4/mod/mod_authz_core.html#require)" directive.
+ Set "`Deny from all`" as well whenever available, to block access regardless of which access control directive takes priority.
+ Assume Apache 2.2 only, if [mod_authz_core](https://httpd.apache.org/docs/2.4/mod/mod_authz_core.html#require) and [mod_access_compat](https://httpd.apache.org/docs/2.4/mod/mod_access_compat.html) are both not available, to avoid doubled directives. In this case set "`Deny from all`" directive only if the providing [mod_authz_host](https://httpd.apache.org/docs/2.2/mod/mod_authz_host.html#deny) module is available. "Satisfy" is a core directive on Apache 2.2.
+ Update Apache version strings. Regarding the used directives/modules, Apache 2.4 and 2.5 (will be released as 2.6) behave the same.
+ Add ordering spaces to better reflect the nested directives and to match style of other `.htaccess` files.
+ <s>Remove one doubled whitespace</s> + some minor syntax "fixes" to match standards:
  - `ifModule` => `IfModule`
  - <s>`$content.= "string";` => `$content .= "string";`</s> **EDIT: Meanwhile done by: https://github.com/nextcloud/server/commit/cdf8c16942c71d0faf71b346beaad70ced22a2fa**

Fixes: https://github.com/nextcloud/server/issues/6449

Signed-off-by: MichaIng <micha@dietpi.com>

MichaIng committed:
Harden data protection .htaccess

+ Set "Satisfy All" whenever available, as well on Apache 2.4+. This is required to override possible "Satisfy Any" on parent dir, which otherwise would allow direct access to data, regardless of "Require" directive.
+ Set "Deny from all" as well whenever available, to block access regardless of which access control directive takes priority.
+ Assume Apache 2.2 only, if mod_authz_core and mod_access_compat are both not available, to avoid doubled directives. In this case set "Deny from all" directive only if the providing mod_authz_host module is available. "Satisfy" is a core directive on Apache 2.2.
+ Update Apache version strings. Regarding the used directives/modules, Apache 2.4 and 2.5 behave the same.
+ Add ordering spaces to better reflect the nested directives and to match style of other .htaccess files.

Fixes: https://github.com/nextcloud/server/issues/6449

Signed-off-by: Micha Felle <micha@dietpi.com>

kesselb commented on lib/private/Setup.php:
```diff
@@ -542,19 +542,27 @@ public static function protectDataDirectory() {
 		//Require all denied
 		$now =  date('Y-m-d H:i:s');
```
```suggestion
		$now = date('Y-m-d H:i:s');
```

MichaIng committed:
Harden config protection .htaccess

+ Set "Satisfy All" whenever available, as well on Apache 2.4+. This is required to override possible "Satisfy Any" on parent dir, which otherwise would allow direct access to data, regardless of "Require" directive.
+ Set "Deny from all" as well whenever available, to block access regardless of which access control directive takes priority.
+ Assume Apache 2.2 only, if mod_authz_core and mod_access_compat are both not available, to avoid doubled directives. In this case set "Deny from all" directive only if the providing mod_authz_host module is available. "Satisfy" is a core directive on Apache 2.2.
+ Update Apache version strings. Regarding the used directives/modules, Apache 2.4 and 2.5 behave the same.
+ Add ordering spaces to better reflect the nested directives and to match style of other .htaccess files.

Fixes: #6449 (for the config directory)

Signed-off-by: Micha Felle <micha@dietpi.com>

kesselb commented on lib/private/Setup.php:
```diff
@@ -542,19 +542,27 @@ public static function protectDataDirectory() {
 		$now =  date('Y-m-d H:i:s');
 		$content = "# Generated by Nextcloud on $now\n";
-		$content.= "# line below if for Apache 2.4\n";
+		$content.= "# Section for Apache 2.4 and 2.5\n";
```
```suggestion
		$content.= "# Section for Apache 2.4 and 2.6\n";
```

http://httpd.apache.org/dev/devnotes.html

MichaIng replied:
@kesselb 
So do I get it right that 2.5 will be finally released as 2.6?
Should be changed in the autoindex part below as well then.

kesselb replied:
I think so

kesselb commented on lib/private/Setup.php:
```diff
@@ -542,19 +542,27 @@ public static function protectDataDirectory() {
 		$content.= "<ifModule mod_authz_core.c>\n";
-		$content.= "Require all denied\n";
+		$content.= "  Require all denied\n";
+		$content.= "</ifModule>\n";
```
```suggestion
		$content .= "</ifModule>\n";
```

There should be a whitespace between $content and `.=`. Mind to change it? I know it was like that before.

MichaIng replied:
@kesselb 
Makes sense, the whitespace is also used throughout other code as well as `ifModule` => `IfModule` you suggested below. Lets keep it consistent and stay with syntax according to official docs, regarding cases.

kesselb commented on lib/private/Setup.php:
```diff
@@ -542,19 +542,27 @@ public static function protectDataDirectory() {
 		$content.= "</ifModule>\n\n";
-		$content.= "# line below if for Apache 2.2\n";
+		$content.= "# Section for Apache 2.2\n";
 		$content.= "<ifModule !mod_authz_core.c>\n";
```
```suggestion
		$content.= "<IfModule !mod_authz_core.c>\n";
```

https://httpd.apache.org/docs/2.4/mod/core.html#ifmodule

MichaIng committed:
Update lib/private/Setup.php

+ Remove unnecessary spaces from code

Co-Authored-By: Daniel Kesselberg <mail@danielkesselberg.de>

MichaIng committed:
Use syntax with cases according to official docs

+ Ref: https://github.com/nextcloud/server/pull/16792/files#r315207691

Signed-off-by: Micha Felle <micha@dietpi.com>

MichaIng committed:
Minor syntax

+ Use Apache syntax with cases according to official docs: https://github.com/nextcloud/server/pull/16792/files#r315207691
+ Add missing whitespace for concatenating strings to variable: https://github.com/nextcloud/server/pull/16792/files#r315207520
+ Apache 2.5 will be released as 2.6: https://github.com/nextcloud/server/pull/16792/files#r315206147

Signed-off-by: Micha Felle <micha@dietpi.com>

MichaIng committed:
Apache 2.5 will be released as 2.6

+ Ref: https://github.com/nextcloud/server/pull/16792/files#r315206147

Signed-off-by: Micha Felle <micha@dietpi.com>

MichaIng commented:
Actually it makes sense to add `Order allow,deny` in front of `Deny from all`. This assures that no `Allow` directive for this dir, declared anywhere else, is evaluated after our `Deny` rule. Otherwise theoretically some faulty or harmful config snipped can override our rule and open access.

MichaIng committed:
Add "Order" to assure that no parental "Allow" can grant access

Signed-off-by: MichaIng <micha@dietpi.com>

MichaIng committed:
Add "Order" to assure that no parental "Allow" can grant access

Signed-off-by: MichaIng <micha@dietpi.com>

MichaIng commented:
@nextcloud/security 
Hey guys, I hope you find some time to look into this and probably we can still merge it with NC17.

I added the `Order Allow,Deny` directive as well. This simply makes sense to assure the `Deny` has always priority. This is as well common across other web applications and their Apache config recommendations and `.htaccess` files to block certain dir access, e.g. [phpBB](https://github.com/phpbb/phpbb/blob/master/phpBB/.htaccess).

nickvergessen commented:
Well 17 is already packaged. But I guess we can do it for 17.0.1

MichaIng committed:
Add missing newlines to .htaccess

Signed-off-by: MichaIng <micha@dietpi.com>

nickvergessen commented:
@rullzer @kesselb time to review so we can merge and get this in?

MichaIng committed:
Merge branch 'master' into patch-1

MichaIng commented:
Resolved conflict caused format fix, which was also included in this PR: https://github.com/nextcloud/server/commit/cdf8c16942c71d0faf71b346beaad70ced22a2fa

gary-kim reviewed (approved):
LGTM
